backend/vkt/src/test/java/fi/oph/vkt/service/PublicReservationServiceTest.java

```diff
@@ -0,0 +1,193 @@
+    );
+    assertEquals(APIExceptionType.CREATE_RESERVATION_CONGESTION, ex.getExceptionType());
+  }
+
+  @Test
+  public void testCreateReservationWithRegistrationClosed() {
+    final ExamEvent examEvent = Factory.examEvent();
+    examEvent.setRegistrationCloses(LocalDate.now().minusDays(1));
+    entityManager.persist(examEvent);
+    final Person person = createPerson();
+
+    final APIException ex = assertThrows(
+      APIException.class,
+      () -> publicReservationService.createReservation(examEvent.getId(), person)
+    );
+    assertEquals(APIExceptionType.CREATE_RESERVATION_REGISTRATION_CLOSED, ex.getExceptionType());
+  }
+
+  private ExamEvent createExamEvent(final int maxParticipants) {
+    final ExamEvent examEvent = Factory.examEvent();
+    examEvent.setMaxParticipants(maxParticipants);
+    entityManager.persist(examEvent);
+
+    return examEvent;
+  }
+
+  private Enrollment createEnrollment(final ExamEvent examEvent, final EnrollmentStatus status) {
+    final Person person = createPerson();
+    final Enrollment enrollment = Factory.enrollment(examEvent, person);
+    enrollment.setStatus(status);
+    entityManager.persist(enrollment);
+
+    return enrollment;
+  }
+
+  private Reservation createReservation(final ExamEvent examEvent, final LocalDateTime expiresAt) {
+    final Person person = createPerson();
+    final Reservation reservation = Factory.reservation(examEvent, person);
+    reservation.setExpiresAt(expiresAt);
+    entityManager.persist(reservation);
+
+    return reservation;
+  }
+
+  private Person createPerson() {
+    final Person person = Factory.person();
+    entityManager.persist(person);
+
+    return person;
+  }
+
+  private void assertReservationDTO(
+    final ExamEvent examEvent,
+    final Person person,
+    final int participants,
+    final PublicReservationDTO dto
+  ) {
+    final PublicExamEventDTO examEventDTO = dto.examEvent();
+    assertEquals(examEvent.getId(), examEventDTO.id());
+    assertEquals(examEvent.getLanguage(), examEventDTO.language());
+    assertEquals(examEvent.getDate(), examEventDTO.date());
+    assertEquals(examEvent.getRegistrationCloses(), examEventDTO.registrationCloses());
+    assertEquals(participants, examEventDTO.participants());
+    assertEquals(examEvent.getMaxParticipants(), examEventDTO.maxParticipants());
+
+    final PublicPersonDTO personDTO = dto.person();
+    assertEquals(person.getId(), personDTO.id());
+    assertEquals(person.getIdentityNumber(), personDTO.identityNumber());
+    assertEquals(person.getLastName(), personDTO.lastName());
+    assertEquals(person.getFirstName(), person.getFirstName());
+    // TODO: fix issue with dto.expiresAt(), in local it's set as 2 hours from UTC,
+    // Using UTC time for tests might be the issue here, while running the app locally the backend returns expiresAt as unix time 30 mins from current time
+    // assertEquals(dto.expiresAt(), System.currentTimeMillis() + ONE_MINUTE.toMillis());
```

Comment:
Jos @terova keksit, miten tän tarkistuksen saa toteuttaa, niin voit ehdottaa tai korjata tähän päälle. Täällä oli aiemmin rivit
```
assertTrue(dto.isValidUntil() > System.currentTimeMillis() + ONE_MINUTE.toMillis() - 20);
assertTrue(dto.isValidUntil() < System.currentTimeMillis() + ONE_MINUTE.toMillis() + 20);
```

Reply:
PublicReservationService kun expiresAt asetetaan, pitäisikö laittaa `ZonedDateTime.of(reservation.getExpiresAt(), ZoneId.of("Europe/Helsinki")).toEpochSecond()` niin silloin `assertEquals(dto.expiresAt(), (System.currentTimeMillis() + ONE_MINUTE.toMillis()) / 1000);` menee läpi.

Reply:
Koska `LocalDateTime.now()` dokumentaatio sanoo että `Returns:
the current date-time using the system clock and default time-zone, not null` niin `.expiresAt(ZonedDateTime.of(reservation.getExpiresAt(), ZoneId.systemDefault()).toEpochSecond())` voisi olla se oikea.